source/agora/common/BlockStorage.d

```diff
@@ -0,0 +1,274 @@
+/*******************************************************************************
+
+    Defines the memory mapped file of a block.
+
+    The number of files is one.
+    We will expand to use multiple files in the next step.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.BlockStorage;
+
+import agora.common.Amount;
+import agora.common.Block;
+import agora.common.Deserializer;
+import agora.common.Serializer;
+
+import std.algorithm.comparison;
+import std.file;
+import std.format;
+import std.mmfile;
+import std.path;
+
+
+/// Ditto
+public class BlockStorage
+{
+    /// Instance of memory mapped file
+    private MmFile file;
+
+    /// Is already mapped ?
+    private bool is_mapped;
+
+    /// Path of block data
+    private string path;
+
+    /// Index is block height, Value is position of file
+    /// We will change to index file in the next step.
+    private size_t[] block_positions;
+
+    /// Ctor
+    public this (string path)
+    {
+        this.is_mapped = false;
+        this.path = path;
+    }
+
+    /***************************************************************************
+
+        Make a file name using the index.
+
+        Returns:
+            Returns the file name.
+
+    ***************************************************************************/
+
+    private string getFileName ()
+    {
+        return buildPath(this.path, "block.dat");
+    }
+
+    /***************************************************************************
+
+        Open memory mapped file.
+
+        If a size change occurs, open it again.
+
+        Params:
+            resize = The size of the file to change. If 0 does not change it.
+
+    ***************************************************************************/
+
+    private void map (size_t resize = 0)
+    {
+        if (this.is_mapped && (resize == 0))
+            return;
+
+        if (this.is_mapped)
+            this.unMap();
+
+        this.file =
+            new MmFile(
+                this.getFileName(),
+                MmFile.Mode.readWrite,
+                resize,
+                null
+            );
+
+        this.is_mapped = true;
+    }
+
+    /***************************************************************************
+
+        Close memory mapped file.
+
+    ***************************************************************************/
+
+    private void unMap ()
+    {
+        if (!this.is_mapped)
+            return;
+        import core.memory : GC;
+
+        destroy(this.file);
+        GC.free(&this.file);
+        this.is_mapped = false;
```

Comment:
Instead of having `is_mapped` boolean, you can just set `this.file = null` and use that as a test for being mapped.

Reply:
Ping